Optimize filter condition with case expression predicate

maswin opened the pull request:
The DB visualization tools such as looker & DbVisualizer generates sub-optimal queries when a selected column has switch case and a filter is applied on top of it. For instance it generates queries such as:

`SELECT case behavior_code when 1 then 'good' when 2 then 'bad' else 'neutral' from students where (case behavior_code when 1 then 'good' when 2 then 'bad' else 'neutral') = 'good';`

But this query can be simplified into:

`SELECT case behavior_code when 1 then 'good' when 2 then 'bad' else 'neutral' from students where behavior_code = 1;`

Without this simplification, the query misses out a lot of other optimizations since the Predicate is not converted to ColumnDomain and not pushed into readers (i.e, ORC indexes) and becomes extremely slow.

Since these are generated queries, the user doesn't have much control on them to convert it manually. 

Internally inside our company we saw lot of such queries running for a long time. So we have come up with this Optimizer to rewrite such case predicates on filter condition. We saw a lot of performance gain from them.

The case expression of format
```
(case 
    when condition1 then result1
    when condition2 then result2
    when condition3 then result3
else elseResult) = value
```
can be converted to a series of AND/OR cases as below:
```
(result1 = value AND condition1)
OR
(result2 = value AND condition2 AND NOT(condition1))
OR
(result3 = value AND condition3 AND NOT(condition1) AND NOT(condition2))
OR
(elseResult = value AND NOT(condition1) AND NOT(condition2) AND NOT(condition3))
```
but this involves a lot of re-evaluation. If we can ensure all the conditions are mutually exclusive then `condition2 AND NOT(condition1)` can be converted to just `condition2` since former being true will mean the later NOT condition is also true since their domains are mutually exclusive. With this the above expression can be further simplified to 
```
(result1 = value AND condition1)
OR
(result2 = value AND condition2)
OR
(result3 = value AND condition3)
OR
(elseResult = value AND NOT(condition1) AND NOT(condition2) AND NOT(condition3))
```
So under the following conditions, we convert case expression predicates to above form:

1. Value is a simple expression(constant/column reference/input reference) and not any function call since it could be re-evaluated.
2. The LHS expression in all WHEN clauses are the same. For example, if one WHEN clause has a condition on col1 and another on col2, it will not work
3. The relational operator in the WHEN clause condition is equals. With other operators it is harder to check for exclusivity.
4. All the RHS expressions are a constant and unique

SimplifyExpressions can discard all conditions in which `result != value` and DomainTranslator can construct a Domain for the column.

martint commented:
@kasiafi, can you take a look at this PR?

ebyhr commented:
@cla-bot check

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+        private Expression getCastExpression(Optional<Cast> castExpression, Expression expression)
+        {
+            return castExpression
+                    .map(x -> (Expression) new Cast(expression, x.getType(), x.isSafe(), x.isTypeOnly()))
```
Rename x to e.g. "cast".

maswin replied:
Done.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+
+            Expression elseResult = defaultValue
+                    .map(x -> (Expression) new ComparisonExpression(operator, getCastExpression(castExpression, x), otherExpression))
+                    .orElse(new IsNullPredicate(otherExpression));
```
This is not correct. We cannot replace any operator with "is null", when one operand is null.
It should rather be:
```
Expression default = defaultValue
        .map(value -> getCastExpression(castExpression, value))
        .orElse(<cast null to the expected type>);

Expression elseResult = new ComparisonExpression(operator, default, otherExpression);

maswin replied:
Made it to return NullLiteral in case of missing else clause since figuring out the expected type from Optimizer is difficult and the Simplify rules applied later adds a CAST if necessary. Now the results are similar to how trino handles other NULL expressions involved comparison function (i.e, NULL > 'string')

I have disabled the unit test with respect to this change - testRewriterOnCaseExpressionWithoutElseClause.
The reason is, the test was failing because of the difference in Cast object created.

When the query is parsed by PlanBuilder.expression, typeOnly is set to false.
```
public Cast(NodeLocation location, Expression expression, DataType type, boolean safe)
{
    this(Optional.of(location), expression, type, safe, false);
}
```
But when the expression is optimized using ExpressionInterpreter, typeOnly is set to true.
```
// Code block from LiteralEncoder.toExpression
if (object == null) {
    if (type.equals(UNKNOWN)) {
        return new NullLiteral();
    }
    return new Cast(new NullLiteral(), toSqlType(type), false, true);
}
```
Not sure if this is an existing bug.

kasiafi replied:
> Not sure if this is an existing bug.

Setting `typeOnly` to false, and not revisiting it later for the expected expression, should be considered a simplification of the testing process rather than a bug. I agree, however, that it might get in the way. This issue was already encountered and addressed here: https://github.com/trinodb/trino/blob/7b0990fcd70b1fe2995caa1071916571b75693eb/core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestSimplifyExpressions.java#L390-L396

maswin replied:
Done.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+                    .map(x -> (Expression) new ComparisonExpression(operator, getCastExpression(castExpression, x), otherExpression))
+                    .orElse(new IsNullPredicate(otherExpression));
+            andExpressions.add(ExpressionUtils.and(elseResult, ExpressionUtils.and(invertedOperands)));
+            ExpressionUtils.and(invertedOperands);
```
A leftover line?

maswin replied:
My bad. This line is unnecessary. Previous line has this expression.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+ * <p>
+ * <p>
+ * This conversion happens under the following conditions:
+ * This conversion happens when all of the following criteria are met:
```
Removed the extra line.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+ * <p>
+ * This conversion happens under the following conditions:
+ * This conversion happens when all of the following criteria are met:
+ * 1. Value is either a constant or column reference or input reference and not any function
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+ * <p>
+ * (result1 = value AND expression=constant1) OR
+ * (result2 = value AND expression=constant2) OR
+ * (elseResult = value AND !condition1 AND !condition2)
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+    @Override
+    public Rule.Result apply(FilterNode filterNode, Captures captures, Rule.Context context)
+    {
+        Expression rewritten = ExpressionTreeRewriter.rewriteWith(new Rewriter(), filterNode.getPredicate(), context);
```
Switched it to Void.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+            return expression instanceof Literal;
+        }
+
+        private Expression processCaseExpression(Optional<Cast> castExpression,
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+            return validateRHSExpressions(rightHandSideExpressions);
+        }
+
+        private boolean validateRHSExpressions(List<Expression> expressions)
```
Done.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,279 @@
+        @Override
+        public Expression rewriteComparisonExpression(ComparisonExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (isCaseExpression(node.getLeft()) && isSimpleExpression(node.getRight())) {
```
Rewrite the sub-expressions recursively.

maswin replied:
Done.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,279 @@
+
+            Expression defaultExpression = defaultValue
+                    .map(value -> getCastExpression(castExpression, value))
+                    .orElse(getCastExpression(castExpression, new NullLiteral()));
```
For consistency with `LiteralEncoder`, cast of `NullLiteral` should be not safe and type-only. Please use the `LiteralEncoder.toExpression()` method to get the proper form of `Cast` over `NullLiteral`.

maswin replied:
Done.

kasiafi commented:
> Made it to return NullLiteral in case of missing else clause since figuring out the expected type from Optimizer is difficult and the Simplify rules applied later adds a CAST if necessary.

It is not correct to insert an expression with mismatching type (here: `UNKNOWN`). We expect that both sides of comparison are of the same type. To determine the type in case when there is no `Cast`, you can use the `TypeAnalyzer.getType()` method. The `TypeAnalyzer` is available for the Optimizer rules in `PlanOptimizers`.

kasiafi reviewed (commented):
Please make sure that there are tests with coercions. E.g. different case results have different types _and_ the compared value type is different. We should check both cases: when the whole `Case` expression is wrapped in cast, and when the compared value is wrapped in cast.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,279 @@
+            return true;
+        }
+
+        private boolean isLiteral(Expression expression)
```
Done

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,279 @@
+
+        private boolean canRewriteSearchedCaseExpression(SearchedCaseExpression caseExpression)
+        {
+            Expression commonLHSExpression = null;
```
Done

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,279 @@
+            }
+            return caseExpression instanceof SimpleCaseExpression ?
+                    processCaseExpression((SimpleCaseExpression) caseExpression, castExpression, otherExpression, operator) :
+                    processCaseExpression((SearchedCaseExpression) caseExpression, castExpression, otherExpression, operator);
```
Done

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,279 @@
+            return expression instanceof SimpleCaseExpression || expression instanceof SearchedCaseExpression;
+        }
+
+        private boolean isSimpleExpression(Expression expression)
```
Done

maswin commented:
> Please make sure that there are tests with coercions. E.g. different case results have different types _and_ the compared value type is different. We should check both cases: when the whole `Case` expression is wrapped in cast, and when the compared value is wrapped in cast.

I have a test unit case "testRewriterOnCaseExpressionWithCastFunction" which tests the whole case expression being wrapped and compared value wrapped in cast cases.
Additionally I see that before the optimizer is reached the common super type of the expression is identified and all other values are wrapped with CAST to bring it all to uniform type. 
Let me know if more test cases are required.

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestRewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    @Test(enabled = false)
+    public void testRewriterOnCaseExpressionWithoutElseClause()
```
We need to get it tested somehow. Have you tried to follow my suggestion https://github.com/trinodb/trino/pull/10580#discussion_r791604255?

maswin replied:
Done.

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestRewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,266 @@
+                .doesNotFire();
+    }
+
+    private void assertRewrittenAndOptimizedExpression(String inputExpressionStr,
```
Please move all the helper methods below the tests, and reformat to either have all arguments on the same line or all arguments on separate lines. Also, please do not abbreviate. I think that `inputExpression`, `expectedRewritten`, `expectedOptimized` are sufficiently clear names.

maswin replied:
Done.

kasiafi commented on testing/trino-testing/src/main/java/io/trino/testing/AbstractTestQueryFramework.java:
```diff
@@ -190,6 +190,11 @@ protected void assertQuery(Session session, @Language("SQL") String actual, @Lan
         QueryAssertions.assertQuery(queryRunner, session, actual, h2QueryRunner, expected, false, false, planAssertion);
     }
 
+    protected void assertQuery(Session actualQuerySession, Session expectedQuerySession, @Language("SQL") String sql)
```
What is `expectedQuerySession`?

maswin replied:
Removed and simplified the test.

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestRewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,266 @@
+        assertThat(optimizedExpression).isEqualTo(expectedOptimizedExpression);
+    }
+
+    private Expression getOptimizedExpression(Expression expression)
```
In fact, a unit test for the rule should be limited to what the rule does. So, the "getOptimized" part does not belong here. Moreover, this optimization does not directly reflect what happens in Trino's optimization process.

If you want to test the final form of the expression, you can do this in `TestLogicalPlanner`.

maswin replied:
Done.

kasiafi reviewed (commented):
Thanks for updating and sorry for the delay.
Here's another batch from me. This time I went through the tests too.
First of all, we need to address the correctness issue with literal uniqueness.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+ * DomainTranslator can construct a Domain for the column
+ */
+public class RewriteCaseExpressionPredicate
+        implements Rule<FilterNode>
```
Done

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+                return isSimpleExpression(((Cast) expression).getExpression());
+            }
+            return expression instanceof Literal ||
+                    expression instanceof FieldReference ||
```
Done

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+ * 2. The LHS expression in all WHEN clauses are the same.
+ *    For example, if one WHEN clause has a condition on col1 and another on col2, it will not work
+ * 3. The relational operator in the WHEN clause condition is equals. With other operators it is harder to check for exclusivity.
+ * 4. All the RHS expressions are a constant and unique
```
I have updated the description with more details.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,277 @@
+        {
+            Set<Expression> literals = new HashSet<>();
+            for (Expression expression : expressions) {
+                if (!isLiteral(expression) || literals.contains(expression)) {
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+ * <p>
+ * This conversion happens when all of the following criteria are met:
+ * 1. Value is either a constant or a symbol
+ * 2. The LHS expression in all WHEN clauses are the same.
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+    public RewriteCaseExpressionPredicate(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer)
+    {
+        this.plannerContext = plannerContext;
+        this.typeAnalyzer = typeAnalyzer;
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+{
+    private static final Pattern<FilterNode> PATTERN = filter();
+    private final PlannerContext plannerContext;
+    private TypeAnalyzer typeAnalyzer;
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+            for (Expression expression : expressions) {
+                if (!isEffectivelyLiteral(plannerContext, session, expression) || literals.contains(expression)) {
+                    return false;
+                }
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+        public Expression rewriteComparisonExpression(ComparisonExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            Expression rewritten = node;
+            if (isCaseExpression(node.getLeft()) && isEffectivelyLiteral(plannerContext, session, node.getRight())) {
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+
+            andExpressions.add(ExpressionUtils.and(elseResult, ExpressionUtils.and(invertedOperands)));
+
+            return Optional.of(ExpressionUtils.or(andExpressions));
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,295 @@
+        private Expression getCastExpression(Optional<Cast> castExpression, Expression expression)
+        {
+            return castExpression
+                    .map(cast -> (Expression) new Cast(expression, cast.getType(), cast.isSafe(), cast.isTypeOnly()))
```
You mean the when case result?

maswin commented:
Sorry for the delay in addressing the comments. Let me know if there are any more changes required and if more test cases are needed.
I raised a PR for this optimizer in Presto and got it merged recently: https://github.com/prestodb/presto/pull/17065

kasiafi reviewed (commented):
@maswin after another glance I noticed that the proposed transformation is not correct. I'm sorry for not catching it earlier. This is because of the semantics of `null` values.
Consider this simple example:
```
trino:tiny> SELECT
         -> (
         ->  CASE
         ->     WHEN col = 1 THEN 'a'
         ->     WHEN col = 2 THEN 'b'
         ->  ELSE 'c'
         ->  END
         -> ) = 'c'
         -> FROM (VALUES null) t(col);
 _col0
-------
 true
(1 row)
```
After the transformation, the result is null. This is because of the "inverted conditions" added to the `else` branch which evaluate to `null`.
Please ignore the following comments until we find an alternative approach without this correctness issue.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java:
```diff
@@ -81,6 +81,7 @@
     private double tableScanNodePartitioningMinBucketToTaskRatio = 0.5;
     private boolean mergeProjectWithValues = true;
     private boolean forceSingleNodeOutput = true;
+    private boolean optimizeCaseExpressionPredicate;
```
Transformed expression at times maybe big or could not be simplified, maybe better to have a toggle.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,95 @@
+ * (result1 = value AND expression=constant1) OR
+ * (result2 = value AND expression=constant2 AND !(expression=constant1)) OR
+ * (result3 = value AND expression=constant3 AND !(expression=constant1) AND !(expression=constant2)) OR
+ * (elseResult = value AND !(expression=constant1) AND !(expression=constant2) AND !(expression=constant3))
```
Added a check such that constant cannot be NULL or NullLiteral (Added unit test for that case).
expression being null is handled in the new form of rewrite.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,95 @@
+                joinExpressionRewrite());
+    }
+
+    private static ExpressionRewriter createRewrite(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer)
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Override
+    public Set<Rule<?>> rules()
```
Removed. It will now be applied to all operators.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,95 @@
+
+    private static ExpressionRewriter createRewrite(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer)
+    {
+        requireNonNull(plannerContext, "metadata is null");
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/CaseExpressionPredicateRewriter.java:
```diff
@@ -0,0 +1,251 @@
+                    return false;
+                }
+
+                Type type = typeAnalyzer.getTypes(session, symbolAllocator.getTypes(), expression).get(NodeRef.of(expression));
```
Done.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/CaseExpressionPredicateRewriter.java:
```diff
@@ -0,0 +1,279 @@
+                    type = typeAnalyzer.getTypes(session, symbolAllocator.getTypes(), expression).get(NodeRef.of(expression));
+                }
+                Object constantExpression = evaluateConstantExpression(expression, type, plannerContext, session, new AllowAllAccessControl(), ImmutableMap.of());
+                if (constantExpression == null || constantExpression instanceof NullLiteral || literals.contains(constantExpression)) {
```
`evaluateConstantExpression` never returns an `Expression`, so you can remove `|| constantExpression instanceof NullLiteral`.

maswin replied:
Done.

kasiafi reviewed (commented):
@maswin 
```(result1 = value AND expression IS NOT NULL AND expression=constant1) OR
(result2 = value AND expression IS NOT NULL AND expression=constant2) OR
(result3 = value AND expression IS NOT NULL AND expression=constant3) OR
(elseResult = value AND ((expression IS NULL) OR (!(expression=constant1) AND !(expression=constant2) AND !(expression=constant3)))```

Adding `expression IS NOT NULL` in the branches 1, 2 and 3 is redundant. They will fail anyway for `expression` being `null` because `expression=constant` returns null. The only thing we really need is the additional condition in the last branch.

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestRewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,200 @@
+    @Test
+    public void testRewriterDoesNotFireOnPredicateWithFunctionCallOnComparisonValue()
+    {
+        assertRewriteDoesNotFire("(case when col1=1 then 'case1' when col2=2 then 'case2' else 'default' end) = upper('case1')");
+        assertRewriteDoesNotFire("(case when col1=1 then 10 when col2=2 then 20 else 30 end) = ceil(col1)");
```
In this test, `col1` and `col2` should be the same.
Anyway, I didn't notice we had the condition you're testing here.

maswin replied:
Done.

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestRewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    @Test
+    public void testRewriterDoesNotFireOnInvalidSearchCaseExpression()
```
These are not "invalid" case expressions. Please rename the method.

maswin replied:
Done.

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestRewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,200 @@
+    @Test
+    public void testRewriterDoesNotFireOnInvalidSearchCaseExpression()
+    {
+        // All LHS expressions are not the same
```
Test all LHS expressions the same but not deterministic.

maswin replied:
Done.

maswin commented on testing/trino-testing/src/main/java/io/trino/testing/AbstractTestQueries.java:
```diff
@@ -502,4 +504,55 @@ public void testFilterPushdownWithAggregation()
     }
+
+    @Test
+    public void testCasePredicateRewrite()
```
Modified to use region and nation table. 
I want the results to be same with and without the optimizer enabled. assertQuery seems to compare the results against another DB and ignores the session completely for creating the expected results.

maswin commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestRewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,200 @@
+    public void testRewriterDoesNotFireOnInvalidSearchCaseExpression()
+    {
+        // All LHS expressions are not the same
+        assertRewriteDoesNotFire("(case when col1=1 then 'case1' when col2=2 then 'case2' else 'default' end) = 'case1'");
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/CaseExpressionPredicateRewriter.java:
```diff
@@ -0,0 +1,279 @@
+        {
+            return caseExpression.getWhenClauses().stream()
+                    .map(x -> ((ComparisonExpression) x.getOperand()).getLeft())
+                    .findFirst().get();
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/CaseExpressionPredicateRewriter.java:
```diff
@@ -0,0 +1,279 @@
+                    return false;
+                }
+                if (type == null) {
+                    type = typeAnalyzer.getTypes(session, symbolAllocator.getTypes(), expression).get(NodeRef.of(expression));
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/CaseExpressionPredicateRewriter.java:
```diff
@@ -0,0 +1,279 @@
+{
+    private CaseExpressionPredicateRewriter() {}
+
+    public static Expression rewrite(Expression expression,
```
Done.

maswin commented on core/trino-main/src/main/java/io/trino/sql/planner/CaseExpressionPredicateRewriter.java:
```diff
@@ -0,0 +1,279 @@
+            Set<Object> literals = new HashSet<>();
+            Type type = null;
+            for (Expression expression : expressions) {
+                if (expression instanceof NullLiteral || !isEffectivelyLiteral(plannerContext, session, expression)) {
```
Done.

maswin commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestRewriteCaseExpressionPredicate.java:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    @Test
+    public void testRewriterOnCaseExpressionWithCastFunction()
```
> @maswin
> 
> ```
> (result1 = value AND expression IS NOT NULL AND expression=constant1) OR
> (result2 = value AND expression IS NOT NULL AND expression=constant2) OR
> (result3 = value AND expression IS NOT NULL AND expression=constant3) OR
> (elseResult = value AND ((expression IS NULL) OR (!(expression=constant1) AND !(expression=constant2) AND !(expression=constant3)))
> ```
> 
> Adding `expression IS NOT NULL` in the branches 1, 2 and 3 is redundant. They will fail anyway for `expression` being `null` because `expression=constant` returns null. The only thing we really need is the additional condition in the last branch.

@kasiafi : The idea of this optimization is most of the 'result=value' expression will evaluate to false during the expression simplification and most of the time only one branch will be present in the end. 
The problem with adding IS NULL check only in last branch is, the last branch might be discarded by ExpressionInterpreter during simplification and NULL check will also get discarded. So we need the NULL check in all branches. 

For example, if we add NULL check only to last branch

```
(CASE
  WHEN col = 1 THEN 'a'
  WHEN col = 2 THEN 'b'
  ELSE 'c'
END) = 'a'
```
will be converted to
```
('a'='a' AND col = 1) OR
('b'='a' AND col = 2) OR
('c'='a' AND ((col IS NULL) OR ((col != 1) AND (col != 2)))
```
after simplification will become
`col = 1`

which will result in NULL=1 and return NULL, but we need 'FALSE' to be returned. By adding IS NOT NULL check, the simplification would be
`col IS NOT NULL and col=1`
which will make sure FALSE is returned instead of NULL.

mosabua commented:
👋 @maswin and @kasiafi .. I assume this PR is still active and you are working towards merge. If that is not the case, please close the PR.

maswin commented:
@kasiafi : Hi, please let me know if more changes are required.

mosabua commented:
@kasiafi and @maswin please let me know if it is a good time to review the docs yet or if this PR still ongoing for a while

mosabua commented:
Closing this one out due to inactivity, but please reopen if you would like to pick this back up.

mosabua commented:
Closing this PR due to inactivity. Anyone interested please feel free to pick it back up and continue work on it here or in a new PR.
